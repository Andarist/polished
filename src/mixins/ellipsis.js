// @flow
import messageHandlers from '../internalHelpers/_messageHandlers'

/**
 * CSS to represent truncated text with an ellipsis.
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *   ...ellipsis('250px')
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${ellipsis('250px')}
 * `
 *
 * // CSS as JS Output
 *
 * div: {
 *   'display': 'inline-block',
 *   'maxWidth': '250px',
 *   'overflow': 'hidden',
 *   'textOverflow': 'ellipsis',
 *   'whiteSpace': 'nowrap',
 *   'wordWrap': 'normal'
 * }
 */

function ellipsis(width?: string = '100%') {
  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    if (
      messageHandlers('mixins/ellipsis.js', {
        // eslint-disable-next-line prefer-rest-params
        arrityCheck: { args: arguments, max: 1 },
        typeChecks: { param: width, type: 'string' },
      })
    ) {
      return {}
    }
  }

  return {
    display: 'inline-block',
    maxWidth: width,
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    wordWrap: 'normal',
  }
}

export default ellipsis
