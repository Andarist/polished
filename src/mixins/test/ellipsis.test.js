// @flow
import ellipsis from '../ellipsis'

describe('ellipsis', () => {
  beforeAll(() => {
    global.console = {
      error: jest.fn(),
      warn: jest.fn(),
    }
  })

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error.mockClear()
    // eslint-disable-next-line no-console
    console.warn.mockClear()
  })

  it('should pass parameter to the value of max-width', () => {
    expect({ ...ellipsis('300px') }).toMatchSnapshot()
  })

  it('should properly add rules when block has existing rules', () => {
    expect({
      background: 'red',
      ...ellipsis('300px'),
    }).toMatchSnapshot()
  })

  it('should default max-width to 100%', () => {
    expect({ ...ellipsis() }).toMatchSnapshot()
  })

  it('should throw an error when called with a non-string value', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    ellipsis(1)
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw a warning when called with more than 1 parameter', () => {
    ellipsis('100%', 2)
    // eslint-disable-next-line no-console
    expect(console.warn.mock.calls).toMatchSnapshot()
  })

  it('should throw an error and a warning when called with multiple non-string values', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    ellipsis(1, 2)
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
    // eslint-disable-next-line no-console
    expect(console.warn.mock.calls).toMatchSnapshot()
  })
})
