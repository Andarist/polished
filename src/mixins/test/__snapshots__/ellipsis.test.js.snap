// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ellipsis should default max-width to 100% 1`] = `
Object {
  "display": "inline-block",
  "maxWidth": "100%",
  "overflow": "hidden",
  "textOverflow": "ellipsis",
  "whiteSpace": "nowrap",
  "wordWrap": "normal",
}
`;

exports[`ellipsis should pass parameter to the value of max-width 1`] = `
Object {
  "display": "inline-block",
  "maxWidth": "300px",
  "overflow": "hidden",
  "textOverflow": "ellipsis",
  "whiteSpace": "nowrap",
  "wordWrap": "normal",
}
`;

exports[`ellipsis should throw a warning when called with more than 1 parameter 1`] = `
Array [
  Array [
    "%c ✨ WARNING ✨ ---- mixins/ellipsis.js --%c

ellipsis %cexpects a maximum of %c1 parameter%c. However, you passed %c2 parameters%c. This additional parameter was ignored.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#ellipsis%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: green",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: goldenrod",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`ellipsis should throw an error and a warning when called with multiple non-string values 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/ellipsis.js --%c

ellipsis %cexpects a %cparameter%c(%cundefined %c:%c cssMeasure%c). However, you passed (%c'1' %c:%c number%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#ellipsis%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`ellipsis should throw an error and a warning when called with multiple non-string values 2`] = `
Array [
  Array [
    "%c ✨ WARNING ✨ ---- mixins/ellipsis.js --%c

ellipsis %cexpects a maximum of %c1 parameter%c. However, you passed %c2 parameters%c. This additional parameter was ignored.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#ellipsis%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: green",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: goldenrod",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`ellipsis should throw an error when called with a non-string value 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/ellipsis.js --%c

ellipsis %cexpects a %cparameter%c(%cundefined %c:%c cssMeasure%c). However, you passed (%c'1' %c:%c number%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#ellipsis%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;
